snakemake/executors/flux.py

```diff
@@ -0,0 +1,242 @@
+
+from snakemake.executors import ClusterExecutor, sleep
+from snakemake.executors.common import format_cli_arg, join_cli_args
+from snakemake.logging import logger
+from snakemake.resources import DefaultResources
+
+# Just import flux once
+try:
+    import flux
+    import flux.job
+except ImportError:
+    flux = None
+
+FluxJob = namedtuple("FluxJob", "job jobname jobid callback error_callback flux_future")
+
+
+class FluxExecutor(ClusterExecutor):
+    """
+    The Flux executor deploys workflows to a flux cluster.
+    """
+
+    def __init__(
+        self,
+        workflow,
+        dag,
+        cores,
+        jobname="snakejob.{name}.{jobid}.sh",
+        printreason=False,
+        quiet=False,
+        printshellcmds=False,
+    ):
+        super().__init__(
+            workflow,
+            dag,
+            None,
+            jobname=jobname,
+            printreason=printreason,
+            quiet=quiet,
+            printshellcmds=printshellcmds,
+            assume_shared_fs=False,
+            max_status_checks_per_second=10,
+        )
+
+        # Attach variables for easy access
+        self.workdir = os.path.realpath(os.path.dirname(self.workflow.persistence.path))
+        self.envvars = list(self.workflow.envvars) or []
+
+        # Quit early if we can't access the flux api
+        self._init_flux()
+
+    def _prepare_job_formatter(self):
+        """
+        A job doesn't have an easy status command - instead we do a job
+        listing with a particular format. See src/cmd/flux-jobs.py#L44
+        in flux-framework/flux-core for more attributes.
+        """
+        jobs_format = (
+            "{id.f58:>12} {username:<8.8} {name:<10.10} {status_abbrev:>2.2} "
+            "{ntasks:>6} {nnodes:>6h} {runtime!F:>8h} {success} {exception.occurred}"
+            "{exception.note} {exception.type} {result} {runtime} {status}"
+            "{ranks:h} {t_remaining} {annotations}"
+        )
+        self.jobs_formatter = flux.job.JobInfoFormat(jobs_format)
+
+        # Note there is no attr for "id", its always returned
+        fields2attrs = {
+            "id.f58": (),
+            "username": ("userid",),
+            "exception.occurred": ("exception_occurred",),
+            "exception.type": ("exception_type",),
+            "exception.note": ("exception_note",),
+            "runtime": ("t_run", "t_cleanup"),
+            "status": ("state", "result"),
+            "status_abbrev": ("state", "result"),
+            "t_remaining": ("expiration", "state", "result"),
+        }
+
+        # Set job attributes we will use later to get job statuses
+        self.job_attrs = set()
+        for field in self.jobs_formatter.fields:
+            if field not in fields2attrs:
+                self.job_attrs.update((field,))
+            else:
+                self.job_attrs.update(fields2attrs[field])
+
+    def _init_flux(self):
+        """
+        Ensure we can import flux in the environment.
+        """
```

Comment:
It looks like this function is only used once, I think it might be marginally clearer if the body of this function was added to `__init__`. Plus I think it's Python best practice to initialize all instance attributes (e.g. `self._fexecutor`) in `__init__`.

Reply:
This function was originally callable and would do the import there, and that design was so a developer could create the class without importing flux (which maybe doesn't work and they are debugging). But with this changed design I agree it doesn't make sense to have separate - I think it might just be my tendency to write things into small functions when possible. I'll push a change to update this.

Reply:
This is fixed.